Add initial free-threading page for the guide

ngoldbaum opened the pull request:
Moves the existing content on free-threading in the migration guide into its own page. Also adds some new content about things we know are going to be issues for some users.

Comments and suggestions are very welcome. I'd really appreciate ideas for illustrative code examples to add, if anyone has any.

ahlinc commented on guide/src/free-threading.md:
There is a `see see` typo.

Icxolu commented on guide/src/free-threading.md:
```diff
@@ -0,0 +1,127 @@
+# Supporting Free-Threaded CPython
+
+CPython 3.13 introduces an experimental build of CPython that does not rely on
+the global interpreter lock for thread safety. As of version 0.23, PyO3 also has
```
```suggestion
the global interpreter lock (GIL) for thread safety. As of version 0.23, PyO3 also has
```
Maybe we want to introduce the acronym for the following section?

Icxolu commented on guide/src/free-threading.md:
Maybe we want to also talk a bit about what PyO3/Rust brings to the table here? Maybe about the "fearless concurrency" point, having `Send`/`Sync` allowing to safely build APIs that can be used in this new context?

Icxolu commented on guide/src/free-threading.md:
This is very valuable information, but also pretty technical. Maybe we should put this in a `<details>` tab and summarize that on a higher level, so that it's a bit easier to grasp for new users. Maybe that "GIL" refers to the interaction with the Python interpreter or something like that.

Icxolu commented on guide/src/free-threading.md:
I'm not quite sure how to interpret the "Rust threads" part. I don't think there is a problem in just spawning Rust thread an doing something different non Python related in the background. It's more about detaching the _current_ thread from it's interaction with the Interpreter and handing back control, if I got that correctly.

Icxolu reviewed (commented):
Sounds good already, added some thoughts that came to mind while reading (from someone who hasn't looked a lot into free-threading yet)

(Nit: Maybe we can use a consistent capitalization of "Python" 🙃)

davidhewitt reviewed (commented):
Thanks, this is already looking great to me! I have a few suggestions too...

ngoldbaum commented:
Thanks for all the comments! I hope the new text addresses the concerns and is clearer.

I'd like to add sections on the critical section and PyMutex wrappers, assuming those get merged, and also maybe another section in the docs about multithreaded programming using PyO3? I definitely found it nontrivial to learn and things like `std::thread::scope` should be pointed at explicitly in the PyO3 docs.

Icxolu reviewed (commented):
Very nice, this reads a lot easier to me.

I guess my main concern is still what terminology we use, when we want to talk about interacting with the interpreter.

I think there are now 3 different wordings in different paragraphs. Sometimes it's still referred to as "GIL", then we have "thread state" and we have "attached to the runtime". I think unifying them makes sense to not overload readers. Personally I like the "attached to the runtime" variant because it's high level enough that someone without much experience can get an idea about whats going on (I think?) and it's independent of the build-mode. We can also add a later (sub)section explaining what "attached to the runtime" means as technical documentation.

ngoldbaum commented on guide/src/free-threading.md:
Thanks for the suggestion, I've re-written this to talk about being attached to an "interpreter runtime" which is also vague but maybe a bit more suggestive than "thread state"?

Icxolu replied:
For me this reads really nicely now ❤️ I think it paints a good, high level picture now with all the necessary information without being overwhelmingly technical 👍  Well done!

ngoldbaum commented on guide/src/free-threading.md:
You're right, it's weaker than that, see https://docs.python.org/3/c-api/init.html#non-python-created-threads.

I've updated this so now the restriction I'm referring to only refers to OS threads that need to interact with Python objects, the rayon example you linked to is messing with rust strings, so no need to be attached to a Python interpreter runtime.

Icxolu replied:
Thanks for the link. That makes a lot of sense. In this case I think we're fine, because we don't allow the `Python` token to cross thread boundaries (`!Send`) and `Python::with_gil` (or `Python::attach` in the future?) will register the current thread correctly, right?

ngoldbaum replied:
Yup, `with_gil` is equivalent to calling `PyGILState_Ensure` and `PyGILStste_Release` before and after some C code that interacts with the C API.

davidhewitt commented on guide/src/free-threading.md:
```diff
@@ -0,0 +1,153 @@
+the GIL is acquired. In the free-threaded build there is no GIL, but the same C
+macros that release or acquire the GIL in the GIL-enabled build instead ask the
+interpreter to attach the thread to the Python runtime, and there can be many
+threads simultaneously attached.
```
I think we should add a sentence here along the lines of

"Similarly, in the GIL-enabled build PyO3 uses the `Python<'py>` type and the `'py` lifetime to signify attachment the global interpreter lock; in the freethreaded build these simply mean the thread is currently attached to the Python interpreter."

ngoldbaum replied:
I incorporated this and edited the surrounding paragraphs a bit. I think it reads better now, thanks for the suggestion.

davidhewitt commented on guide/src/free-threading.md:
```diff
@@ -0,0 +1,153 @@
+does not rely on the [global interpreter
+lock](https://docs.python.org/3/glossary.html#term-global-interpreter-lock)
+(often referred to as the GIL) for thread safety. As of version 0.23, PyO3 also
+has preliminary support for building rust extensions for the free-threaded
```
Should we be capitalizing Rust? 🤔 (A few cases below also, if you agree...)

```suggestion
has preliminary support for building Rust extensions for the free-threaded
```

davidhewitt commented on guide/src/free-threading.md:
```diff
@@ -0,0 +1,153 @@
+
+The main reason for attaching to the Python runtime is to interact with Python
+objects or call into the CPython C API. To interact with the Python runtime, the
+thread must register itself by attaching to the interpreter runtime.
```
Maybe mention `with_gil` here as the way to attach and the plan is that it will continue to exist under a new name?

davidhewitt commented on guide/src/free-threading.md:
```diff
@@ -0,0 +1,153 @@
+    // stand-in for something that executes arbitrary Python code
+    let d = PyDict::new(py);
+    d.set_item(PyNone::get(py), PyNone::get(py)).unwrap();
+    // we're not executing Python code while holding the lock, so GILProtected
+    // was never needed
```
```suggestion
    // as with any `Mutex` usage, lock the mutex for as little time as possible
    // in this case, we do it just while pushing into the `Vec`
```

davidhewitt commented on guide/src/free-threading.md:
```diff
@@ -0,0 +1,153 @@
+builds and mutexes otherwise. If your use of `GILProtected` does not guard the
+execution of arbitrary Python code or use of the CPython C API, then conditional
+compilation is likely unnecessary since `GILProtected` was not needed in the
+first place and instead rust mutexes or atomics should be preferred. Python 3.13
+introduces `PyMutex`, which releases the GIL while the lock is held, so that is
+another option if you only need to support newer Python versions.
```
Is this interaction between `PyMutex` and the GIL true? I thought it just released the GIL while blocking.

```suggestion
first place and instead rust mutexes or atomics should be preferred. Python 3.13
introduces `PyMutex`, which releases the GIL while the waiting for the lock, so that is
another option if you only need to support newer Python versions.
```

One option is that we add `MutexExt` traits to `pyo3::sync` which work for all Python versions, similar to #4513 

e.g. `.lock_py_attached(py)`

🤔

ngoldbaum replied:
Thanks for spotting the error, I applied your correction.

The extension trait is an interesting idea, it's certainly possible for PyO3 to detach from the GIL if a mutex is blocked. Definitely something to consider...

davidhewitt reviewed (commented):
Looking great, thanks! Just a few smaller thoughts, and then let's merge 👍

alex commented on guide/src/free-threading.md:
```diff
@@ -0,0 +1,165 @@
+simultaneously on multiple OS threads.  This means it's much more
+straightforward to achieve multithreaded parallelism in the Python layer, using
+the `threading` module, without any need to ensure the GIL is released during
+compute-heavy tasks, or any worries about [Amdahl's
+law](https://en.wikipedia.org/wiki/Amdahl%27s_law) limiting parallel
+performance. If you have ever needed to use `multiprocessing` to achieve a
```
I don't understand the amdahl's law reference here, it's absolutely still relevant :-)

I think what you're trying to say is "in a GIL world, the value of releasing the GIL is bound by the portion of your code that drops the GIL (under Amdahl's law), while in the no GIL world, you can gain teh benefits to the full extent your code is parallel", but that's not clear from the text.

ngoldbaum replied:
I rewrote this to try to make the point about Amdahl's law more explicit. Thanks for pointing out that this was unclear.

alex replied:
Looks good

alex commented on guide/src/free-threading.md:
```diff
@@ -0,0 +1,165 @@
+access to Python state, so if more than one thread tries to access a Python
+object that has already been mutably borrowed, only runtime checking enforces
+safety around mutably aliased Rust variables the Python interpreter can
+access. We believe that it would require adding an `unsafe impl` for `Send` or
+`Sync` to trigger this behavior in code using PyO3. Please report any issues
+related to runtime borrow checker errors on mutable pyclass implementations that
+do not make strong assumptions about the GIL.
```
Sorry, why do we think that `unsafe impl` is required?

ngoldbaum replied:
I'm not actually sure, that came from @davidhewitt

davidhewitt replied:
I think the point is that PyO3 uses `Send` and `Sync` bounds to rely on thread safety, but this is not really different from any other Rust code and also is relevant for everything, not just `#[pyclass]`.

Maybe we can replace this whole structure with something simpler like:

"`#[pyclass]` data is protected from concurrent access by a `RefCell`-like pattern of runtime borrow checking. Like a `RefCell`, PyO3 will raise exceptions (or in some cases panic) to enforce exclusive access for mutable borrows. With the GIL enabled it was unusual for these errors to be encountered in well structured code (using `allow_threads` or calling a `&self` `#[pymethod]` from a `&mut self` one could produce these errors)."

alex replied:
Yes, that's much clearer. Perhaps we could even include an example of a Python snippet likely to fail? `[threading.Thread(target=some_obj.some_method_that_takes_self_mut).start() for _ in range(16)]` (but cleaned up)

ngoldbaum replied:
I added a code example.

One question: do the code examples in the docs get run as doctests? I modeled what I added based on another example in the guide that has some python code, but looking at the nox configuration I don't see anything that checks the code examples in the guide.

davidhewitt replied:
That's a great question - you should add this new `.md` file to the list of `doctests!` in the bottom of `lib.rs` (around line 500).

alex commented on guide/src/free-threading.md:
```diff
@@ -0,0 +1,165 @@
+`GILProtected` is a PyO3 type that allows mutable access to static data by
+leveraging the GIL to lock concurrent access from other threads. In
+free-threaded Python there is no GIL, so you will need to replace this type with
+some other form of locking. In many cases, a type from `std::sync::Atomic` or
```
```suggestion
some other form of locking. In many cases, a type from `std::sync::atomic` or
```

davidhewitt reviewed (approved):
Looks great, thanks!
